heron/common/src/java/com/twitter/heron/common/network/HeronServerTester.java

```diff
@@ -0,0 +1,367 @@
+//  Copyright 2017 Twitter. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+package com.twitter.heron.common.network;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+import com.google.protobuf.GeneratedMessage;
+import com.google.protobuf.Message;
+
+import com.twitter.heron.api.metric.MultiCountMetric;
+import com.twitter.heron.common.basics.ByteAmount;
+import com.twitter.heron.common.basics.Communicator;
+import com.twitter.heron.common.basics.NIOLooper;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+/**
+ * Class to help simplify testing HeronServer implementations by consolidating common test
+ * functionality. A new instance of this class should be used for each test. After initializing the
+ * class, tests should call start() which guarantees the following:
+ * <ol>
+ * <li>the server has started</li>
+ * <li>the client has been launched</li>
+ * </ol>
+ * Upon test completion, the stop() method should be closed to properly release resources. This
+ * method is safe to be called on an instance that was never properly started.
+ */
+public class HeronServerTester {
```

Comment:
Yeah good call @nlu90 . I moved `TestCommunicator` into `c.t.h.common.basics` to be along side `Communicator`. I also moved `awaitMultiCountMetric` into the test that needs it (not yet put up for PR).

The remaining inner classes are now more closely coupled with `HeronServerTester`. They're basically the handlers and the client helper classes. I'm normally not that big a fan of inner classes but in this case I think it makes sense to keep them as inner classes for now since a.) their impl is closely coupled to `HeronServerTester` and, b.) I don't want to pollute the `c.t.h.common.network` package space with too many test-related classes.